src/Http/Routing/src/EndpointDataSource.cs

```diff
@@ -22,4 +24,148 @@ public abstract class EndpointDataSource
     /// Returns a read-only collection of <see cref="Endpoint"/> instances.
     /// </summary>
     public abstract IReadOnlyList<Endpoint> Endpoints { get; }
+
+    /// <summary>
+    /// Get the <see cref="RouteEndpoint"/> instances for this <see cref="EndpointDataSource"/> given the specified <see cref="RouteGroupContext.Prefix"/>.
+    /// </summary>
+    /// <param name="context">Details about how the returned <see cref="RouteEndpoint"/> instances should be grouped and a reference to application services.</param>
+    /// <returns>
+    /// Returns a read-only collection of <see cref="RouteEndpoint"/> instances given the specified group <see cref="RouteGroupContext.Prefix"/> and <see cref="RouteGroupContext.Conventions"/>.
+    /// </returns>
+    public virtual IReadOnlyList<Endpoint> GetEndpointGroup(RouteGroupContext context)
+    {
+        // Only evaluate Endpoints once per call.
+        var endpoints = Endpoints;
```

Comment:
The dependency between `Endpoints` and `GetEndpointGroup` is a little scary. If you build a data source that implements Endpoints by calling `GetEndpointGroup`, it'll explode.